sycl/source/device_selector.cpp

```diff
@@ -89,24 +105,38 @@ device device_selector::select_device() const {
   }
 
   if (res != nullptr) {
-    if (detail::pi::trace(detail::pi::TraceLevel::PI_TRACE_BASIC)) {
-      std::string PlatformName = res->get_info<info::device::platform>()
-                                     .get_info<info::platform::name>();
-      std::string DeviceName = res->get_info<info::device::name>();
-      std::cout << "SYCL_PI_TRACE[all]: "
-                << "Selected device ->" << std::endl
-                << "SYCL_PI_TRACE[all]: "
-                << "  platform: " << PlatformName << std::endl
-                << "SYCL_PI_TRACE[all]: "
-                << "  device: " << DeviceName << std::endl;
-    }
+    traceDeviceSelection(*res, score, true);
+
     return *res;
   }
 
   throw cl::sycl::runtime_error("No device of requested type available.",
                                 PI_ERROR_DEVICE_NOT_FOUND);
 }
 
+// select_device(selector)
+__SYCL_EXPORT device
+select_device(DSelectorInvocableType DeviceSelectorInvocable) {
+  std::vector<device> Devices = device::get_devices();
+
+  return select_device(DeviceSelectorInvocable, Devices);
+}
+
+// select_device(selector, context)
+__SYCL_EXPORT device
+select_device(DSelectorInvocableType DeviceSelectorInvocable,
+              const context &SyclContext) {
+  std::vector<device> devices = SyclContext.get_devices();
+
+  return select_device(DeviceSelectorInvocable, devices);
+}
+
+} // namespace detail
+
+device device_selector::select_device() const {
```

Comment:
Do you need to implement the old SYCL 1.2.1 ugly Java-factory style API ?

Reply:
I'm no fan either, but this is still used if compiling with C++14.  Just preserving the functionality that was here before.